### PR TITLE
Implement particle filtering for neuland simulation

### PR DESCRIPTION
**Pull request by SimonL2718**
## New features:

Allow filtering for certain particles during digitization. Setup is done within neulandAna.cxx executable.

---

Checklist:

* [x] Rebased against `dev` branch
* [x] My name is in the resp. CONTRIBUTORS/AUTHORS file
* [x] Followed the [pull request guidlines](https://opensource.creativecommons.org/contributing-code/pr-guidelines/) and the [Git workflow](https://github.com/AnarManafov/GitWorkflow/blob/master/GitWorkflow.markdown)
* [x] Followed the [seven rules of great commit messages](https://chris.beams.io/posts/git-commit/#seven-rules)

### CODE REVIEW

**Comment by YanzhaoW** [comment]
Hi @jose-luis-rs . @SimonL2718 is a bachelor student from my group. Could you add him to the contributors such that the CI workflow could run?

File: CMakeLists.txt [hunk 1]
```diff
@@ -549,6 +549,7 @@ if(NOT MODULE)
     add_subdirectory(tracking)
     add_subdirectory(passive)
     add_subdirectory(tcal)
+    add_subdirectory(r3bgen)
```
Comment (YanzhaoW) [inline CMakeLists.txt]: Maybe don't change it here.

File: neuland/digitizing/NeulandPointFilter.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+} // namespace R3B::Neuland
+class NeulandPointFilter
+{
+  public:
+    NeulandPointFilter() = default;
+    void SetFilter(R3B::Neuland::BitSetParticle filtered_particles);
+    void SetFilter(R3B::Neuland::BitSetParticle filtered_particles, double minimum_allowed_energy);
+    [[nodiscard]] auto GetFilter() const -> R3B::Neuland::BitSetParticle { return filtered_particles_; }
+    [[nodiscard]] auto GetMinimumAllowedEnergy() const -> double { return minimum_allowed_energy_; }
+    auto ShouldNeulandPointBeFiltered(const R3BNeulandPoint& neuland_point) -> bool;
```
Comment (YanzhaoW) [inline neuland/digitizing/NeulandPointFilter.h]: The function name is too long. Your input parameter already contains `NeulandPoint` type and you don't need to mention it again in the function name.

```suggestion
    auto CheckFiltered(const R3BNeulandPoint& neuland_point) -> bool;
```

File: neuland/digitizing/NeulandPointFilter.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+  private:
+    R3B::Neuland::BitSetParticle filtered_particles_ = R3B::Neuland::BitSetParticle::none;
+    // energy in GeV
+    double minimum_allowed_energy_ = 0;
```
Comment (YanzhaoW) [inline neuland/digitizing/NeulandPointFilter.h]: Put the comments in the same line.

```suggestion
    double minimum_allowed_energy_ = 0; // energy in GeV
```

File: neuland/executables/neulandSim.cxx [hunk 4]
```diff
@@ -22,7 +22,10 @@ constexpr int DEFAULT_RUNID = 999;
 int main(int argc, const char** argv)
 {
     auto timer = TStopwatch{};
-    auto const PID = 2112;
+    auto const neutron_PID = 2112;
+    auto const Sn_p = int{ 50 };
+    auto const Sn_z = int{ 123 };
+    auto const BeamEnergyAtTarget = 883.;
```
Comment (YanzhaoW) [inline neuland/executables/neulandSim.cxx]:
```suggestion
    constexpr auto neutron_PID = 2112;
    constexpr auto Sn_p = 50;
    constexpr auto Sn_z = 123;
    constexpr auto BeamEnergyAtTarget = 883.;
```

File: neuland/simulation/R3BNeuland.cxx [hunk 5]
```diff
@@ -23,27 +23,29 @@
...
+#include <fmt/format.h>
 
 // Initialize variables from Birk' s Law
-static constexpr Double_t BirkdP = 1.032;
-static constexpr Double_t BirkC1 = 0.013 / BirkdP;
-static constexpr Double_t BirkC2 = 9.6e-6 / (BirkdP * BirkdP);
+constexpr auto seconds_to_nanoseconds = 1e9;
+static constexpr double BirkdP = 1.032;
+static constexpr double BirkC1 = 0.013 / BirkdP;
+static constexpr double BirkC2 = 9.6e-6 / (BirkdP * BirkdP);
```
Comment (YanzhaoW) [inline neuland/simulation/R3BNeuland.cxx]:
```suggestion
constexpr auto seconds_to_nanoseconds = 1e9;
constexpr auto BirkdP = 1.032;
constexpr auto BirkC1 = 0.013 / BirkdP;
constexpr auto BirkC2 = 9.6e-6 / (BirkdP * BirkdP);
```

File: neuland/simulation/R3BNeuland.cxx [hunk 6]
```diff
@@ -23,27 +23,29 @@
...
 
-inline Double_t GetLightYield(const Int_t charge, const Double_t length, const Double_t edep)
+inline auto GetLightYield(const int charge, const double length, const double edep) -> double
 {
     // Apply Birk's law ( Adapted from G3BIRK/Geant3)
     if (charge != 0 && length > 0)
     {
-        Double_t birkC1Mod = BirkC1;
+        double birkC1Mod = BirkC1;
```
Comment (YanzhaoW) [inline neuland/simulation/R3BNeuland.cxx]:
```suggestion
        constexpr auto birkC1Mod = BirkC1;
```

File: r3bdata/neulandData/R3BNeulandPoint.h [hunk 7]
```diff
@@ -37,32 +39,44 @@ class R3BNeulandPoint : public FairMCPoint
...
 
-    R3BNeulandPoint(const FairMCPoint& point, const Double_t lightYield)
+    R3BNeulandPoint(const FairMCPoint& point,
+                    const Double_t lightYield,
+                    const int particle_id,
+                    const int parent_particle_id)
         : FairMCPoint(point)
         , fLightYield(lightYield)
+        , particle_id_(particle_id)
+        , parent_particle_id_(parent_particle_id)
```
Comment (YanzhaoW) [inline r3bdata/neulandData/R3BNeulandPoint.h]: Use curly brackets:

```suggestion
        : FairMCPoint { point }
        , fLightYield{ lightYield }
        , particle_id_{ particle_id }
        , parent_particle_id_{ parent_particle_id}
```

**Review by YanzhaoW — COMMENTED** [review]
Hi, Simon

I leave some comments. Please fix them.

File: CMakeLists.txt (hunk 1 above)
Comment (SimonL2718) [reply CMakeLists.txt]: Should it be with the entries for r3bbase and r3bdata or somewhere else entirely?

File: r3bdata/neulandData/R3BNeulandPoint.h (hunk 7 above)
Comment (SimonL2718) [reply r3bdata/neulandData/R3BNeulandPoint.h]: Should I change this for other construction methods of NeulandPoints aswell or just this one in particular?

File: CMakeLists.txt (hunk 1 above)
Comment (YanzhaoW) [reply CMakeLists.txt]: Hi, just revert back to the dev version. (between music and lumon).

File: r3bdata/neulandData/R3BNeulandPoint.h (hunk 7 above)
Comment (YanzhaoW) [reply r3bdata/neulandData/R3BNeulandPoint.h]: Yes, better change the other construction methods of neuland points.

File: r3bdata/neulandData/R3BNeulandPoint.h [hunk 8]
```diff
@@ -37,32 +39,44 @@ class R3BNeulandPoint : public FairMCPoint
                     const Double_t length,
                     const Double_t eLoss,
                     const UInt_t EventId,
-                    const Double_t lightYield)
-        : FairMCPoint(trackID, detID, pos, mom, tof, length, eLoss, EventId)
-        , fLightYield(lightYield)
+                    const Double_t lightYield,
+                    const int particle_id,
+                    const int parent_particle_id)
```
Comment (YanzhaoW) [inline r3bdata/neulandData/R3BNeulandPoint.h]: We could just tell static analyser to ignore this warning by adding `NOLINTBEGIN` and `NOLINTEND`
```c++
// NOLINTBEGIN
R3BNeulandPoint(const Int_t trackID,
                    const Int_t detID,
                    const TVector3& pos,
                    const TVector3& mom,
                    const Double_t tof,
                    const Double_t length,
                    const Double_t eLoss,
                    const UInt_t EventId,
                    const Double_t lightYield,
                    const int particle_id,
                    const int parent_particle_id)
// NOLINTEND
```

File: neuland/test/digitizing/testNeulandPointFilter.cxx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    // Test for NeulandPointFilter::ShouldNeulandPointBeFiltered function
+    TEST(NeulandPointFilterTest, ShouldNeulandPointBeFilteredTest)
+    {
+        constexpr auto proton_pid = 2212;
+        constexpr auto neutron_pid = 2112;
+        constexpr auto eLoss_proton = 0.7;
+        constexpr auto eLoss_neutron = 0.3;
+        TVector3 sample_vector;
```
Comment (YanzhaoW) [inline neuland/test/digitizing/testNeulandPointFilter.cxx]:
```suggestion
        const auto sample_vector = TVector3{};
```

NOTE: you should always use following style to initialize a variable:
```c++
auto var_name = Type {}; // const auto if necessary
```

File: neuland/test/digitizing/testNeulandPointFilter.cxx [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+        NeulandPointFilter filter;
+
+        R3BNeulandPoint protonPoint(0, 0, sample_vector, sample_vector, 0., 0., eLoss_proton, 0, 0., proton_pid, 0);
```
Comment (YanzhaoW) [inline neuland/test/digitizing/testNeulandPointFilter.cxx]:
```suggestion
        const auto protonPoint = R3BNeulandPoint {0, 0, sample_vector, sample_vector, 0., 0., eLoss_proton, 0, 0., proton_pid, 0};
```

File: neuland/test/digitizing/testNeulandPointFilter.cxx [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        R3BNeulandPoint neutronPoint(0, 0, sample_vector, sample_vector, 0., 0., eLoss_neutron, 0, 0., neutron_pid, 0);
```
Comment (YanzhaoW) [inline neuland/test/digitizing/testNeulandPointFilter.cxx]:
```suggestion
        const auto neutronPoint = R3BNeulandPoint {0, 0, sample_vector, sample_vector, 0., 0., eLoss_neutron, 0, 0., neutron_pid, 0};
```

**Review by YanzhaoW — COMMENTED** [review]
Here are some tips to fix the warnings from Clang-tidy

**Comment by YanzhaoW** [comment]
To fix the clang-format issue, just click its "Details", find the file names, go there and use "Space f m" to autoformat the file in neovim.

File: r3bdata/neulandData/R3BNeulandPoint.h [hunk 12]
```diff
@@ -25,9 +25,9 @@ class R3BNeulandPoint : public FairMCPoint
   public:
     R3BNeulandPoint()
         : FairMCPoint{}
```
Comment (YanzhaoW) [inline r3bdata/neulandData/R3BNeulandPoint.h]:
```suggestion
    R3BNeulandPoint()
```

File: r3bdata/neulandData/R3BNeulandPoint.h [hunk 13]
```diff
@@ -42,21 +42,21 @@ class R3BNeulandPoint : public FairMCPoint
                     const Double_t lightYield,
```
Comment (YanzhaoW) [inline r3bdata/neulandData/R3BNeulandPoint.h:41]: Just disable this warning by:

```c++
// NOLINTBEGIN
R3BNeulandPoint(const Int_t trackID,
                    const Int_t detID,
                    const TVector3& pos,
                    const TVector3& mom,
                    const Double_t tof,
                    const Double_t length,
                    const Double_t eLoss,
                    const UInt_t EventId,
                    const int particle_id,
                    const int parent_particle_id)
// NOLINTEND
```

**Comment by jose-luis-rs** [comment]
Hi, this branch was squashed in #1084

**Comment by jose-luis-rs** [comment]
Please @SimonL2718 reopen this PR. It should be approved by the NeuLAND WG before merging.

**Comment by YanzhaoW** [comment]
Please see this #1086 for the further discussion.

**Pull request merged**